messaging/publisher/publisher.go

```diff
@@ -0,0 +1,455 @@
+func (p *Publisher) GetOurInstallations() ([]*multidevice.Installation, error) {
+	if p.protocol == nil {
+		return nil, errProtocolNotInitialized
+	}
+
+	privateKeyID := p.whisper.SelectedKeyPairID()
+	if privateKeyID == "" {
+		return nil, errNoKeySelected
+	}
+
+	privateKey, err := p.whisper.GetPrivateKey(privateKeyID)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.protocol.GetOurInstallations(&privateKey.PublicKey)
+}
+
+// SetInstallationMetadata sets the metadata for our own installation
+func (p *Publisher) SetInstallationMetadata(installationID string, data *multidevice.InstallationMetadata) error {
+	if p.protocol == nil {
+		return errProtocolNotInitialized
+	}
+
+	privateKeyID := p.whisper.SelectedKeyPairID()
+	if privateKeyID == "" {
+		return errNoKeySelected
+	}
+
+	privateKey, err := p.whisper.GetPrivateKey(privateKeyID)
+	if err != nil {
+		return err
+	}
+
+	return p.protocol.SetInstallationMetadata(&privateKey.PublicKey, installationID, data)
+}
+
+func (p *Publisher) GetPublicBundle(identityKey *ecdsa.PublicKey) (*protobuf.Bundle, error) {
+	if p.protocol == nil {
+		return nil, errProtocolNotInitialized
+	}
+
+	return p.protocol.GetPublicBundle(identityKey)
+}
+
+func (p *Publisher) Start(online func() bool, startTicker bool) error {
+	p.online = online
+	if startTicker {
+		p.startTicker()
+	}
+	return nil
+}
+
+func (p *Publisher) Stop() error {
+	if p.filter != nil {
+		if err := p.filter.Stop(); err != nil {
+			log.Error("Failed to stop filter service with error", "err", err)
+		}
+	}
+
+	return nil
+}
+
+func (p *Publisher) getNegotiatedChat(identity *ecdsa.PublicKey) *filter.Chat {
+	return p.filter.GetNegotiated(identity)
+}
+
+func (p *Publisher) LoadFilters(chats []*filter.Chat) ([]*filter.Chat, error) {
+	return p.filter.Init(chats)
+}
+
+func (p *Publisher) LoadFilter(chat *filter.Chat) ([]*filter.Chat, error) {
+	return p.filter.Load(chat)
+}
+
+func (p *Publisher) RemoveFilters(chats []*filter.Chat) error {
+	return p.filter.Remove(chats)
+}
+
+func (p *Publisher) ProcessMessage(msg *whisper.Message, msgID []byte) error {
+	if !p.config.PFSEnabled {
+		return ErrPFSNotEnabled
+	}
+
+	privateKeyID := p.whisper.SelectedKeyPairID()
+	if privateKeyID == "" {
+		return errNoKeySelected
+	}
+
+	privateKey, err := p.whisper.GetPrivateKey(privateKeyID)
+	if err != nil {
+		return err
+	}
+
+	publicKey, err := crypto.UnmarshalPubkey(msg.Sig)
+	if err != nil {
+		return err
+	}
+
+	// Unmarshal message
+	protocolMessage := &protobuf.ProtocolMessage{}
+
+	if err := proto.Unmarshal(msg.Payload, protocolMessage); err != nil {
+		p.log.Debug("Not a protocol message", "err", err)
+		return ErrNoProtocolMessage
+	}
+
+	response, err := p.protocol.HandleMessage(privateKey, publicKey, protocolMessage, msgID)
+	switch err {
+	case nil:
+		msg.Payload = response
+	case chat.ErrDeviceNotFound:
+		// TODO(adam): why is this condition needed? What's the different if it it's false?
```

Comment:
@cammellos can you please help me with this one? And two lines below, I don't understand what it means if public keys are different in this case.

Reply:
`ErrDeviceNotFound` is thrown when we receive a message that is not targeted to our device.

Our own devices are manage explicitly, meaning that if we see a new device with our own public key (say A1 & A2), we will not be targeting `A2` by default, it needs to be explicitly enabled by the user (by calling `EnableInstallation(deviceID)`. That's for security reason, as we expect the user to be in control of his own devices, and he knows which one to trust (so if a user steals our own key, we won't be disclosing our own data to the new device).
That means that if we received a message from our own public key, not for our device, it's because the user *explicitly* did not want our device to receive it (because it's not a paired device).

When it comes to devices we don't own (say B1 and B2), those are managed implicitly, as there's no source of truth, so, A1 will send to B1 & B2 even if they are not paired, as we don't really know which one to trust.
So if we receive a message from B that is not targeting our device, we want to let B know that we are also to be targeted (which is what this signal does, upon receiving a message that is not targeted to us, we will reply with our bundle, so we are included in the next message).

Not sure if it's clear

Reply:
Also, this signal will not need to be sent to the client (any client) anymore, eventually this can all be handled in the backend (`status-go` or whatever), basically upon seeing that we are not targeted, we just reply with a message, and we send maximum a message every x hours (we don't want to reply to any message, otherwise it's too much traffic)

Reply:
Ok, so for the case where the public keys (from the message and our) are the same, first case, we can return `ErrNotPairedDevice` which probably gonna be only logged and skipped.

For the second case, maybe `ErrInvalidDeviceTarget` can be returned and in such a case, the signal will be sent.

Does it sound legitimate? I would even put this logic into `HandleMessage` to return a well specified error.

Reply:
yes, sounds much better!

Reply:
Actually, the second case is covered:
```go
        msg := msgs[s.config.InstallationID]
	if msg == nil {
		msg = msgs[noInstallationID]
	}

	// We should not be sending a signal if it's coming from us, as we receive our own messages
	if msg == nil && *theirIdentityKey != myIdentityKey.PublicKey {
		return nil, ErrDeviceNotFound
	}
```

I also realised that the first option is not gonna happen because as you said we don't send a message to not paired device. However, if some other implementation did that, we will have a panic because the code above is followed by `payload := msg.GetPayload()` and `msg` being `nil`.

I will cover only this case with `ErrNotPairedDevice `.

Reply:
> I also realised that the first option is not gonna happen because as you said we don't send a message to not paired device

It actually does happen because of the way whisper works, essentially we use the whisper pk to target messages, so anyone with your pk will receive the message:
```
{
  installation-id-1: encrypted-payload1
  installation-id-2: encrypted-payload2
}
```

So any of your installations will receive this, but only `installation-1` and `installation-2` will be able to decrypt the content, while `installation-3` will receive the message, but will have no content that can be decrypted

Reply:
Ah right, I did it correctly (https://github.com/status-im/status-go/pull/1508/commits/5bfcdbe9c8a325eac3bb65614df15c7144a0f786#diff-e155aad09e72a37bdf6f38432d5b9ab6R241) but described it in a wrong way. If a message is sent from device 1 or 2, of course device 3 will receive it but the installation won't match. Without the change mentioned above it could panic or this case is handled somewhere else earlier?

Reply:
Interesting, it definitely does not panic as we would have noticed, but it's not handled, so just by looking at the code it should panic.

would it be possible that this always return true? ` *theirIdentityKey != myIdentityKey.PublicKey` is using underneath `BigInt` and possibly is comparing pointers instead (as I understand it should use `Cmp`). Status react also makes the same check so we never noticed.

Reply:
Ah, that might be a case. This is never true... In the console client, we check if explicitly with `Cmp` to avoid that: https://github.com/status-im/status-console-client/blob/6009b73feb651d0df793f758c1e630b99af4b051/protocol/client/stream.go#L81. I will fix it.